File: src/me/shkschneider/skeleton/security/SimpleCrypt.java
package me.shkschneider.skeleton.security;

import me.shkschneider.skeleton.helpers.LogHelper;
import me.shkschneider.skeleton.java.StringHelper;

@SuppressWarnings("unused")
public final class SimpleCrypt {

    private static final String ALGORITHM = "XOR";

    private char[] mKey;

    public SimpleCrypt(final String key) {
        if (StringHelper.nullOrEmpty(key)) {
            LogHelper.debug("Key was NULL");
            mKey = "".toCharArray();
            return ;
        }

        mKey = key.toCharArray();
    }

    public String algorithm() {
        return ALGORITHM;
    }

    public String key() {
        return new String(mKey);
    }

    public String encrypt(final String string) {
        if (StringHelper.nullOrEmpty(string)) {
            LogHelper.warning("String was NULL");
            return null;
        }

        final char[] chars = string.toCharArray();
        final char[] encrypted = new char[chars.length];
        for (int i = 0; i < chars.length; i++) {
            encrypted[i] = (char) (chars[i] ^ mKey[i % mKey.length]);
        }
        return Base64Helper.encrypt(new String(encrypted).getBytes());
    }

    public String decrypt(final String string) {
        if (StringHelper.nullOrEmpty(string)) {
            LogHelper.warning("String was NULL");
            return null;
        }

        char[] encrypted = Base64Helper.decrypt(string).toCharArray();
        char[] decrypted = new char[encrypted.length];
        for (int i = 0; i < encrypted.length; i++) {
            decrypted[i] = (char) (encrypted[i] ^ mKey[i % mKey.length]);
        }
        return new String(decrypted);
    }

}

File: kotlin/src/main/kotlin/me/shkschneider/skeleton/kotlin/security/SimpleCrypt.kt
package me.shkschneider.skeleton.kotlin.security

import me.shkschneider.skeleton.kotlin.util.fromBase64

private const val ALGORITHM = "XOR"

/**
 * > Welcome to the missile launch web interface!
 * > Enter the target's coordinates.
 * > Enter your email address for our records.
 * > Enter you email again, to ensure you typed it correctly.
 * <https://xkcd.com/970/>
 */
open class SimpleCrypt(key: String) : ICrypt<String>(key) {

    private val key: CharArray = key.toCharArray()

    override fun algorithm(): String =
        ALGORITHM

    override fun key(): String {
        return String(key)
    }

    override fun encrypt(src: String): String? {
        val chars = src.toCharArray()
        val encrypted = CharArray(chars.size)
        for (i in chars.indices) {
            encrypted[i] = (chars[i].toInt() xor key[i % key.size].toInt()).toChar()
        }
        return String(encrypted) // TODO test
    }

    override fun decrypt(src: String): String? =
        src.fromBase64().let { encrypted -> // TODO test
            val decrypted = CharArray(encrypted.size)
            for (i in encrypted.indices) {
                decrypted[i] = (encrypted[i].toInt() xor key[i % key.size].toInt()).toChar()
            }
            return String(decrypted)
        }

}
